### PR TITLE
integrate mesos cloud provider with k8s proper

### PR DESCRIPTION
**Pull request by jdef**
Mesos cloud provider implementation. see #6676 

cc @davidopp @bgrant0607 @guenter

### CODE REVIEW

**Comment by jdef** [comment]
I signed it!

I'm working for mesosphere and they've signed the CLA.

**Comment by jdef** [comment]
xref https://github.com/mesosphere/kubernetes-mesos/issues/114

**Comment by bgrant0607** [comment]
cc @roberthbailey

**Comment by errordeveloper** [comment]
@jdef I can see that `TCPLoadBalancer` is not implemented here. Reading #6676 it seems as if the intention is that eventually one can chain providers together, am I understanding it correctly with respect to external LB functionality?

**Comment by jdef** [comment]
Yes, that's the goal. This initial implementation is pretty vanilla. Future
iterations would add support to layer GCE, AWS, e.g. cloud support so that
we could leverage load balancers, etc.

On Tue, May 12, 2015 at 7:18 AM, Ilya Dmitrichenko <notifications@github.com

**Comment by errordeveloper** [comment]
> Future iterations would add support to layer GCE, AWS, e.g. cloud support so that we could leverage load balancers, etc.

@jdef that's great, is there a proposal for this? #2770 seems related, but hasn't seen much activity...

**Comment by errordeveloper** [comment]
Also, I think there need to be some more explanation of how exactly Kubernetes on Mesos works and what's the benefit of doing so. I have attempted to find out, but it's hard to tell from the code. All I have  understood is that launching Kubernetes on a Mesos cluster magically makes all Mesos slaves available to Kubernetes, but how does it work is still mystery to me.

**Comment by jdef** [comment]
There was some discussion around this a while back:
https://github.com/GoogleCloudPlatform/kubernetes/pull/5980

We followed up with a hangout session to discuss and decided that keeping
things simple for now was a good bet, and that later iterations could
expand on a simple implementation.

On Tue, May 12, 2015 at 8:45 AM, Ilya Dmitrichenko <notifications@github.com

**Comment by jdef** [comment]
Does that explanation need to be a part of this PR? We're planning future
PR's to upstream the custom scheduler and executor components.

The diagrams are *slightly* out of date but perhaps this is a good starting
point? https://github.com/mesosphere/kubernetes-mesos/tree/master/docs


On Tue, May 12, 2015 at 8:47 AM, Ilya Dmitrichenko <notifications@github.com

**Comment by errordeveloper** [comment]
> Does that explanation need to be a part of this PR?

No, thanks for the link!

File: pkg/cloudprovider/mesos/client.go [hunk 1]
```diff
@@ -0,0 +1,296 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
Comment (roberthbailey) [inline pkg/cloudprovider/mesos/client.go]: nit: 2015 for new files

File: pkg/cloudprovider/mesos/client.go [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	log "github.com/golang/glog"
+	"github.com/mesos/mesos-go/detector"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	"golang.org/x/net/context"
+)
+
+const (
```
Comment (roberthbailey) [inline pkg/cloudprovider/mesos/client.go]: nit: if there's just one const/var defined it doesn't need to be put into a block

File: pkg/cloudprovider/mesos/client.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+type stateCache struct {
+	sync.Mutex
+	expiresAt time.Time
+	cached    *mesosState
+	err       error
+	ttl       time.Duration
+	refill    func(context.Context) (*mesosState, error)
+}
+
+// Reload the state cache if it has expired.
```
Comment (roberthbailey) [inline pkg/cloudprovider/mesos/client.go]: nit: function descriptions should start with the name of the function, e.g. "reloadCache reloads the state cache if it has expired."

File: pkg/cloudprovider/mesos/client.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+}
+
+// Returns the cached Mesos state.
+func (c *stateCache) getCachedState(ctx context.Context) (*mesosState, error) {
+	c.reloadCache(ctx)
+	return c.cached, c.err
+}
+
+// Returns the cached Mesos cluster name.
+func (c *stateCache) getClusterName(ctx context.Context) (string, error) {
```
Comment (roberthbailey) [inline pkg/cloudprovider/mesos/client.go]: nit: "getter" functions in go don't start with "get". This should be renamed to "clusterName"

**Comment by roberthbailey** [comment]
@jdef generally we ask that you separate out godeps changes into a separate commit to make your code changes easier to review. I'm not sure if it's possible to back it out now that they are squashed together, but it's something to keep in mind as you add new dependencies.

**Comment by jdef** [comment]
@roberthbailey I attempted to address your code concerns and also split up the commit. HTH

**Comment by roberthbailey** [comment]
Thanks. They were really just some drive-by comments, leaving a more thorough review to @davidopp.

File: pkg/cloudprovider/mesos/mesos.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+// ListClusters lists the names of the available Mesos clusters.
+func (c *MesosCloud) ListClusters() ([]string, error) {
+	// Always returns a single cluster (this one!)
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	name, err := c.client.clusterName(ctx)
+	return []string{name}, err
+}
+
+// Master gets back the address (either DNS name or IP address) of the master node for the cluster.
```
Comment (davidopp) [inline pkg/cloudprovider/mesos/mesos.go]: Clarify that this is the Mesos master, not the Kubernetes master.

File: pkg/cloudprovider/mesos/mesos.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+			}
+
+			return host, nil
+		}
+	}
+	return "", errors.New(fmt.Sprintf("The supplied cluster '%v' does not exist", clusterName))
+}
+
+// ipAddress returns an IP address of the specified instance.
+func (c *MesosCloud) ipAddress(name string) (net.IP, error) {
```
Comment (davidopp) [inline pkg/cloudprovider/mesos/mesos.go]: Why is "c" a method receiver for this function -- it doesn't use "c" anywhere AFAICT.

**Comment by theiq** [comment]
Hi there, i am very curious about the mesos integration in kubernetes. What is the state of the integration and is there some kine of use-guide how to get it up and running? By the way, i sthe kubernetes-mesos project dead and it it succeeded by the kubernetes integration or will it still receive some updates?

**Comment by jdef** [comment]
@theiq the kubernetes-mesos project is not dead and is the current
reference implementation of the integration.

plans are moving forward to upstream most of (if not all of) the code from
that project into k8s, as per #6676. this PR is part of that effort.

**Comment by theiq** [comment]
i don't know where or whom to ask, so sorry in advance. (when) will the kubernets-mesos project support version 0.17 or anything which understands API v1beta3 (v1)?

**Comment by jdef** [comment]
v1beta3 support: soon-ish. sorry, but don't have an exact date for you.

File: pkg/cloudprovider/mesos/mesos.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+		log.V(2).Infof("failed to resolve IP from host name '%v': %v", name, err)
+		return nil, err
+	}
+	ipaddr = iplist[0]
+	log.V(2).Infof("resolved host '%v' to '%v'", name, ipaddr)
+	return ipaddr, nil
+}
+
+// ExternalID returns the cloud provider ID of the specified instance.
+func (c *MesosCloud) ExternalID(instance string) (string, error) {
```
Comment (davidopp) [inline pkg/cloudprovider/mesos/mesos.go]: this one doesn't need to receive MesosCloud either

File: pkg/cloudprovider/mesos/mesos.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+				return node.resources, nil
+			}
+		}
+	}
+	log.Warningf("failed to locate node spec for %q", name)
+	return nil, nil
+}
+
+// NodeAddresses returns the addresses of the specified instance.
+func (c *MesosCloud) NodeAddresses(name string) ([]api.NodeAddress, error) {
```
Comment (davidopp) [inline pkg/cloudprovider/mesos/mesos.go]: same here

File: pkg/cloudprovider/mesos/mesos.go (hunk 7 above)
Comment (jdef) [inline pkg/cloudprovider/mesos/mesos.go]: actually *MesosCloud needs to be the receiver to implement
cloudprovider.Instances


On Thu, May 14, 2015 at 5:10 AM, David Oppenheimer <notifications@github.com

File: pkg/cloudprovider/mesos/mesos.go (hunk 8 above)
Comment (jdef) [inline pkg/cloudprovider/mesos/mesos.go]: actually *MesosCloud needs to be the receiver to implement
cloudprovider.Instances

On Thu, May 14, 2015 at 5:12 AM, David Oppenheimer <notifications@github.com

**Comment by davidopp** [comment]
LGTM

**Pull request merged**